### PR TITLE
CDK: #3764 obfuscate secrets in logging

### LINKED ISSUE #3764: CDK: implement custom logging formatter to obfuscate secrets
## Tell us about the problem you're trying to solve
Currently, we have multiple ways to expose credentials in logs (info, exceptions, tests)

## Describe the solution you’d like
The only real way to secure credentials is to control logging output with a logging Formatter attached to the root logger.
This handler will get the connector's config and will replace all occurrence of values from fields marked as secret.

```python
class SecureFormatter(logging.Formatter):
    """ Custom formatter to prevent secrets leaking
    """
    def format(self, record: 'logging.LogRecord') -> str:
        msg = super().format(record)
        for secret in secrets:
            msg = msg.replace(secret, '***')
        return msg
```

## Describe the alternative you’ve considered or used
do manually

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200405283415880) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by eliziario**
## What
Implements #3764 - obsfuscate airbyte secrets from logging output

## How
A LoggingFilter that reads the secret property names from the spec and gets the corresponding values from the connector config to be replace by a placeholder. 

## Recommended reading order
1. `entrypoint.py`
2. `test_secure_logger.py`
3. `secure_logger.py`

## 🚨 User Impact 🚨
cdk connectors will need to be version bumped in order to take advantage of this new functionality

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/secure_logger.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def filter(self, record: logging.LogRecord) -> str:
+        record.msg = reduce(
+            lambda log_msg, secret: log_msg.replace(secret, "****"),
+            self.secrets_strings,
+            record.msg,
+        )
+        return True
+
+    @staticmethod
+    def get_secrets(source, config):
```
Comment (keu) [inline airbyte-cdk/python/airbyte_cdk/secure_logger.py:45]: this will not work in case of nested secrets

Comment (eliziario) [reply airbyte-cdk/python/airbyte_cdk/secure_logger.py:45]: Humm, ok, let me fix that

File: airbyte-cdk/python/airbyte_cdk/secure_logger.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+from .logger import LOGGING_CONFIG, TRACE_LEVEL_NUM, AirbyteNativeLogger
+
+
+def init_secure_logger(name: str = None, source: Optional[Source] = None, config: Optional[Mapping] = None):
+    """Initial set up of logger"""
+    logging.setLoggerClass(AirbyteNativeLogger)
+    logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+    logger = logging.getLogger(name)
+    logger.propagate = True
```
Comment (eliziario) [inline airbyte-cdk/python/airbyte_cdk/secure_logger.py:20]: note to self: remove it

**Comment by eliziario** [comment]
Eugene, could not find a way to capture all output that wouldn't interfere with our normal use of stdout. Added unhandled exception capturing

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py [hunk 3]
```diff
@@ -5,26 +5,27 @@
...
+from airbyte_cdk.utils.airbyte_secrets_utils import get_secrets
 
 logger = init_logger("airbyte")
 
 
 class AirbyteEntrypoint(object):
     def __init__(self, source: Source):
         self.source = source
-        self.logger = logging.getLogger(f"airbyte.{getattr(source, 'name', '')}")
+        self.logger = init_logger(f"airbyte.{getattr(source, 'name', '')}")
```
Comment (keu) [inline airbyte-cdk/python/airbyte_cdk/entrypoint.py:25]: but in this case, we will have it initialized twice

File: airbyte-cdk/python/airbyte_cdk/entrypoint.py [hunk 4]
```diff
@@ -106,7 +113,7 @@ def launch(source: Source, args: List[str]):
     source_entrypoint = AirbyteEntrypoint(source)
     parsed_args = source_entrypoint.parse_args(args)
     for message in source_entrypoint.run(parsed_args):
-        print(message)
+        print(message + "\n")
```
Comment (keu) [inline airbyte-cdk/python/airbyte_cdk/entrypoint.py:116]:
```suggestion
        print(message)
```

File: airbyte-cdk/python/airbyte_cdk/logger.py [hunk 5]
```diff
@@ -29,19 +32,39 @@
 }
 
 
+def init_unhandled_exception_output_filtering(logger: logging.Logger) -> None:
+    def hook_fn(_logger, exception_type, exception_value, traceback_):
```
Comment (keu) [inline airbyte-cdk/python/airbyte_cdk/logger.py:36]: docstring would be nice here

**Review by keu — APPROVED** [review]
minor fixes left

**Review by keu — APPROVED** [review]

**Pull request merged**